runtime/near-vm-logic/src/logic.rs

```diff
@@ -1147,26 +1154,35 @@ impl<'a> VMLogic<'a> {
         pub_key_len: u64,
         pub_key_ptr: u64,
     ) -> Result<u64> {
-        use ed25519_dalek::{PublicKey, Signature, Verifier, SIGNATURE_LENGTH};
+        use ed25519_dalek::{PublicKey, Signature, Verifier, PUBLIC_KEY_LENGTH, SIGNATURE_LENGTH};
 
         self.gas_counter.pay_base(ed25519_verify_base)?;
-        if sig_len != SIGNATURE_LENGTH as u64 {
+        let msg = self.get_vec_from_memory_or_register(msg_ptr, msg_len)?;
+        let signature_array = self.get_vec_from_memory_or_register(sig_ptr, sig_len)?;
+        if signature_array.len() != SIGNATURE_LENGTH {
             return Err(VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput {
                 msg: "invalid signature length".to_string(),
             }));
         }
-        let msg = self.get_vec_from_memory_or_register(msg_ptr, msg_len)?;
-        let signature_array = self.get_vec_from_memory_or_register(sig_ptr, sig_len)?;
-        let signature = Signature::from_bytes(&signature_array).map_err(|e| {
-            VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput { msg: e.to_string() })
-        })?;
+
+        let signature = match Signature::from_bytes(&signature_array) {
+            Ok(signature) => signature,
+            Err(_) => return Ok(0),
+        };
         let num_bytes = msg.len();
         self.gas_counter.pay_per(ed25519_verify_byte, num_bytes as _)?;
 
         let pub_key_array = self.get_vec_from_memory_or_register(pub_key_ptr, pub_key_len)?;
-        let pub_key = PublicKey::from_bytes(&pub_key_array).map_err(|e| {
-            VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput { msg: e.to_string() })
-        })?;
+        if pub_key_array.len() != PUBLIC_KEY_LENGTH {
+            return Err(VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput {
+                msg: "invalid public key length".to_string(),
+            }));
+        }
+        let pub_key = match PublicKey::from_bytes(&pub_key_array) {
+            Ok(pub_key) => pub_key,
+            Err(_) => return Ok(0),
+        };
+
```

Comment:
Ok, this now is correct I think. One think I am not sure about is in which order should we fetch various things. I suggest to maybe stick with `sig,msg,key` order, so something like this:

```
        self.gas_counter.pay_base(ed25519_verify_base)?;

        let signature_array = self.get_vec_from_memory_or_register(sig_ptr, sig_len)?;
        if signature_array.len() != SIGNATURE_LENGTH {
            return Err(VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput {
                msg: "invalid signature length".to_string(),
            }));
        }

        let signature = match Signature::from_bytes(&signature_array) {
            Ok(signature) => signature,
            Err(_) => return Ok(0),
        };

        let msg = self.get_vec_from_memory_or_register(msg_ptr, msg_len)?;
        let num_bytes = msg.len();
        self.gas_counter.pay_per(ed25519_verify_byte, num_bytes as _)?;

        let pub_key_array = self.get_vec_from_memory_or_register(pub_key_ptr, pub_key_len)?;
        if pub_key_array.len() != PUBLIC_KEY_LENGTH {
            return Err(VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput {
                msg: "invalid public key length".to_string(),
            }));
        }
        let pub_key = match PublicKey::from_bytes(&pub_key_array) {
            Ok(pub_key) => pub_key,
            Err(_) => return Ok(0),
        };

        match pub_key.verify(&msg, &signature) {
            Err(_) => Ok(0),
            Ok(()) => Ok(1),
        }

```


But I don't have a great intuition here as to what's the right choice. 

The curretn code, were we first fetch message,then signature, then charge for messge, does seem odd though.